deployment/postgres-init-db/sql/tables/scheduler/scheduling_specification/scheduling_specification_goals.sql

```diff
@@ -98,34 +98,31 @@ create function scheduler.update_scheduling_specification_goal_func()
   returns trigger
   language plpgsql as $$
   declare
-    next_priority integer;
+    max_priority integer;
 begin
-  select coalesce(
-    (select priority
-     from scheduler.scheduling_specification_goals ssg
-     where ssg.specification_id = new.specification_id
-     order by priority desc
-     limit 1), -1) + 1
-  into next_priority;
+  select count(goal_invocation_id) - 1
+  from scheduler.scheduling_specification_goals ssg
+  where ssg.specification_id = new.specification_id
+  into max_priority;
```

Comment:
Looking at the name of the variable (`max_priority`), would this perhaps be a more direct way to express this query?

```suggestion
  select max(priority)
  from scheduler.scheduling_specification_goals ssg
  where ssg.specification_id = new.specification_id
  into max_priority;
```

Reply:
Yeah, it is. I went with the `count` approach initially, as I wanted to bound `priority` not by its own value but by a property of the specification: that is: the `max_permitted_priority` (I dropped `permitted` when naming the variable) is the number of constraints/goals on the spec (minus 1 b/c priority is 0-based). I was also testing this against a DB where I had already let `priority` escape this upper bound, which meant I couldn't use it as the bounding column.

Reply:
That's a good point - I poked around on aerie-dev and found cases where the bounds had been escaped. The questions that come to mind are:
1. Do we feel like we understand how that happened?
2. Is it a problem?
3. How does this PR change that story? Will existing scheduling specifications be affected?

Reply:
1. Yes, [I've explained it how it happens in the UI PR](https://github.com/NASA-AMMOS/aerie-ui/pull/1537#discussion_r2001567862)
2. It is not a problem with execution AFAIK
3. Currently, this PR isn't cleaning up existing specs (they're allowed to exist but will have issues changing the priorities unless they explicit set them to be back in-bounds), but I can add a step to the migration to fix the existing specs (keep the relative order but bring them back in-bounds)

Reply:
> Currently, this PR isn't cleaning up existing specs (they're allowed to exist but will have issues changing the priorities unless they explicit set them to be back in-bounds), but I can add a step to the migration to fix the existing specs (keep the relative order but bring them back in-bounds)

Got it - is it easy for a user to fix this issue? It seems like it could be rather disruptive if a user is prevented from doing their work due to our "bugfix".

Any of the following would be okay:
- A migration automatically fixes the priorities on existing specifications
- Existing specifications autofix when a user changes the priority of a goal
- We continue to permit specifications that don't strictly go from 0 to n - 1 (or is it 1 to n? I'm not sure)

Reply:
I've done the first option, and "compressed" the priorities back in bounds during the migration (plan version shared below):

```sql
with priorities as (
  select
    goal_invocation_id,
    row_number() over (partition by specification_id) as new_prio
  from scheduler.scheduling_specification_goals ssg
  order by priority
)
update scheduler.scheduling_specification_goals ssg
set priority = p.new_prio - 1 -- -1, as priority starts at 0
from priorities p
where p.goal_invocation_id = ssg.goal_invocation_id;
```